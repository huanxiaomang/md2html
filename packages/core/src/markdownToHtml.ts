// 正则表达式常量
const HEADER_REGEX = /^(#{1,6}) /;
const UNORDERED_LIST_ITEM_REGEX = /^- /;
const ORDERED_LIST_ITEM_REGEX = /^\d+\. /;
const HORIZONTAL_RULE_REGEX = /^---$/;
const BLOCKQUOTE_REGEX = /^> /;
const INLINE_CODE_REGEX = /`(.*?)`/g;
const STRONG_TEXT_REGEX = /\*\*(.*?)\*\*/g;
const EMPHASIS_TEXT_REGEX = /__(.*?)__/g;
const EMPHASIS_TEXT_ALT_REGEX = /\*(.*?)\*/g;
const EMPHASIS_TEXT_ALT2_REGEX = /_(.*?)_/g;
const IMAGE_TAG_REGEX = /!\[(.*?)\]\((.*?)\)/g;
const LINK_TAG_REGEX = /\[(.*?)\]\((.*?)\)/g;

// HTML 标签转换函数
const wrapInTag = (tag: string, content: string): string => `<${tag}>${content}</${tag}>`;

// 处理行内容
const processLine = (line: string, context: {
    inCodeBlock: boolean;
    inBlockquote: boolean;
    listStack: string[];
    html: string;
}): void => {
    if (context.inCodeBlock) {
        if (line.startsWith('```')) {
            context.html += '</code></pre>';
            context.inCodeBlock = false;
        } else {
            context.html += `${line}\n`;
        }
        return;
    }

    if (line.startsWith('```')) {
        closeLists(context);
        closeBlockquote(context);
        context.html += '<pre><code>';
        context.inCodeBlock = true;
        return;
    }

    if (BLOCKQUOTE_REGEX.test(line)) {
        if (!context.inBlockquote) {
            closeLists(context);
            context.html += '<blockquote>';
            context.inBlockquote = true;
        }
        line = line.replace(BLOCKQUOTE_REGEX, '');
    } else if (context.inBlockquote && !BLOCKQUOTE_REGEX.test(line)) {
        closeBlockquote(context);
    }

    const headerMatch = HEADER_REGEX.exec(line);
    if (headerMatch) {
        closeLists(context);
        closeBlockquote(context);
        const headerLevel = headerMatch[1].length;
        const headerText = line.substring(headerMatch[0].length);
        context.html += wrapInTag(`h${headerLevel}`, headerText);
        return;
    }

    if (UNORDERED_LIST_ITEM_REGEX.test(line)) {
        if (!context.listStack.includes('ul')) {
            context.html += '<ul>';
            context.listStack.push('ul');
        }
        line = line.replace(UNORDERED_LIST_ITEM_REGEX, '');
        context.html += wrapInTag('li', line);
        return;
    }

    if (ORDERED_LIST_ITEM_REGEX.test(line)) {
        if (!context.listStack.includes('ol')) {
            context.html += '<ol>';
            context.listStack.push('ol');
        }
        line = line.replace(ORDERED_LIST_ITEM_REGEX, '');
        context.html += wrapInTag('li', line);
        return;
    }

    if (context.listStack.length && !ORDERED_LIST_ITEM_REGEX.test(line) && !UNORDERED_LIST_ITEM_REGEX.test(line)) {
        closeLists(context);
    }

    if (HORIZONTAL_RULE_REGEX.test(line)) {
        context.html += '<hr />';
        return;
    }

    line = line
        .replace(IMAGE_TAG_REGEX, '<img src="$2" alt="$1"/>')
        .replace(LINK_TAG_REGEX, '<a href="$2">$1</a>')
        .replace(STRONG_TEXT_REGEX, '<strong>$1</strong>')
        .replace(EMPHASIS_TEXT_REGEX, '<strong>$1</strong>')
        .replace(EMPHASIS_TEXT_ALT_REGEX, '<em>$1</em>')
        .replace(EMPHASIS_TEXT_ALT2_REGEX, '<em>$1</em>')
        .replace(INLINE_CODE_REGEX, '<code>$1</code>');

    if (line.trim()) {
        context.html += wrapInTag('p', line);
    }
};

// 关闭所有列表标签
const closeLists = (context: { listStack: string[]; html: string }) => {
    while (context.listStack.length) {
        const tag = context.listStack.pop();
        context.html += `</${tag}>`;
    }
};

// 关闭引用块
const closeBlockquote = (context: { inBlockquote: boolean; html: string }) => {
    if (context.inBlockquote) {
        context.html += '</blockquote>';
        context.inBlockquote = false;
    }
};

// 主转换函数
export function markdownToHtml(markdown: string): string {
    const lines = markdown.split('\n');
    const context = {
        listStack: [] as string[],
        inBlockquote: false,
        inCodeBlock: false,
        html: ''
    };

    lines.forEach(line => processLine(line, context));

    closeBlockquote(context);
    if (context.inCodeBlock) {
        context.html += '</code></pre>';
    }
    closeLists(context);

    return context.html;
}
