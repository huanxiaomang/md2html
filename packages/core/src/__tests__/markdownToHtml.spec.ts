import { describe, expect, it } from 'vitest';
import { markdownToHtml } from '../markdownToHtml';

describe('markdownToHtml', () => {
    it('should convert a single line of plain text', () => {
        const markdown = 'Just some plain text.';
        const expectedHtml = '<p>Just some plain text.</p>';
        expect(markdownToHtml(markdown)).toBe(expectedHtml);
    });

    it('should convert multiple paragraphs of plain text', () => {
        const markdown = 'First paragraph.\n\nSecond paragraph.';
        const expectedHtml = '<p>First paragraph.</p><p>Second paragraph.</p>';
        expect(markdownToHtml(markdown)).toBe(expectedHtml);
    });

    it('should handle text with no formatting', () => {
        const markdown = 'This is a simple text with no special Markdown syntax.';
        const expectedHtml = '<p>This is a simple text with no special Markdown syntax.</p>';
        expect(markdownToHtml(markdown)).toBe(expectedHtml);
    });

    it('should convert headers without nested elements', () => {
        const markdown = '# Header 1\n## Header 2\n### Header 3';
        const expectedHtml = '<h1>Header 1</h1><h2>Header 2</h2><h3>Header 3</h3>';
        expect(markdownToHtml(markdown)).toBe(expectedHtml);
    });

    it('should convert bold text', () => {
        const markdown = '**Bold Text**';
        const expectedHtml = '<p><strong>Bold Text</strong></p>';
        expect(markdownToHtml(markdown)).toBe(expectedHtml);
    });

    it('should convert italic text', () => {
        const markdown = '*Italic Text*';
        const expectedHtml = '<p><em>Italic Text</em></p>';
        expect(markdownToHtml(markdown)).toBe(expectedHtml);
    });

    it('should convert inline code', () => {
        const markdown = '`console.log()`';
        const expectedHtml = '<p><code>console.log()</code></p>';
        expect(markdownToHtml(markdown)).toBe(expectedHtml);
    });

    it('should convert code blocks', () => {
        const markdown = '```\nfunction test() {\n  return true;\n}\n```';
        const expectedHtml = '<pre><code>function test() {\n  return true;\n}\n</code></pre>';
        expect(markdownToHtml(markdown)).toBe(expectedHtml);
    });

    it('should convert links without nested elements', () => {
        const markdown = '[hxm](http://czstudio.tech)';
        const expectedHtml = '<p><a href="http://czstudio.tech">hxm</a></p>';
        expect(markdownToHtml(markdown)).toBe(expectedHtml);
    });

    it('should convert images without nested elements', () => {
        const markdown = '![Alt text](https://img.bosszhipin.com/static/file/2024/mctx1cdg891707037634775.png)';
        const expectedHtml = '<p><img src="https://img.bosszhipin.com/static/file/2024/mctx1cdg891707037634775.png" alt="Alt text"/></p>';
        expect(markdownToHtml(markdown)).toBe(expectedHtml);
    });

    it('should convert ordered lists without nested elements', () => {
        const markdown = '1. First item\n2. Second item\n3. Third item';
        const expectedHtml = '<ol><li>First item</li><li>Second item</li><li>Third item</li></ol>';
        expect(markdownToHtml(markdown)).toBe(expectedHtml);
    });

    it('should convert unordered lists without nested elements', () => {
        const markdown = '- First item\n- Second item\n- Third item';
        const expectedHtml = '<ul><li>First item</li><li>Second item</li><li>Third item</li></ul>';
        expect(markdownToHtml(markdown)).toBe(expectedHtml);
    });

    it('should handle blockquotes without nested elements', () => {
        const markdown = '> This is a blockquote.';
        const expectedHtml = '<blockquote><p>This is a blockquote.</p></blockquote>';
        expect(markdownToHtml(markdown)).toBe(expectedHtml);
    });

    it('should handle horizontal rules', () => {
        const markdown = '---';
        const expectedHtml = '<hr />';
        expect(markdownToHtml(markdown)).toBe(expectedHtml);
    });

    it('should handle empty input', () => {
        const markdown = '';
        const expectedHtml = '';
        expect(markdownToHtml(markdown)).toBe(expectedHtml);
    });

    it('should handle mixed content', () => {
        const markdown = '# Header\n\n**Bold** and *italic* text.\n\n- List item 1\n- List item 2\n\n`inline code`';
        const expectedHtml = '<h1>Header</h1><p><strong>Bold</strong> and <em>italic</em> text.</p><ul><li>List item 1</li><li>List item 2</li></ul><p><code>inline code</code></p>';
        expect(markdownToHtml(markdown)).toBe(expectedHtml);
    });

    it('should handle multiple code blocks', () => {
        const markdown = '```\nCode block 1\n```\n\n```\nCode block 2\n```';
        const expectedHtml = '<pre><code>Code block 1\n</code></pre><pre><code>Code block 2\n</code></pre>';
        expect(markdownToHtml(markdown)).toBe(expectedHtml);
    });

    it('should handle mixed lists', () => {
        const markdown = '- Unordered item 1\n- Unordered item 2\n\n1. Ordered item 1\n2. Ordered item 2';
        const expectedHtml = '<ul><li>Unordered item 1</li><li>Unordered item 2</li></ul><ol><li>Ordered item 1</li><li>Ordered item 2</li></ol>';
        expect(markdownToHtml(markdown)).toBe(expectedHtml);
    });

    it('should close open lists and blockquotes before new content', () => {
        const markdown = '- List item\n> Blockquote\nSome other text';
        const expectedHtml = '<ul><li>List item</li></ul><blockquote><p>Blockquote</p></blockquote><p>Some other text</p>';
        expect(markdownToHtml(markdown)).toBe(expectedHtml);
    });

    it('should handle closing open lists and blockquotes at end of input', () => {
        const markdown = '- List item\n> Blockquote';
        const expectedHtml = '<ul><li>List item</li></ul><blockquote><p>Blockquote</p></blockquote>';
        expect(markdownToHtml(markdown)).toBe(expectedHtml);
    });

    it('should close a code block at the end of input', () => {
        const markdown = '```\nCode block\n```';
        const expectedHtml = '<pre><code>Code block\n</code></pre>';
        expect(markdownToHtml(markdown)).toBe(expectedHtml);
    });

    it('should handle an open code block with no closing backticks', () => {
        const markdown = '```\nOpen code block';
        const expectedHtml = '<pre><code>Open code block\n</code></pre>'; // Ensure the code block is closed properly
        expect(markdownToHtml(markdown)).toBe(expectedHtml);
    });

    it('should match snapshot for mixed content', () => {
        const markdown = '# Header\n\n**Bold** and *italic* text.\n\n- List item 1\n- List item 2\n\n`inline code`';
        expect(markdownToHtml(markdown)).toMatchSnapshot();
    });

    it('should match snapshot for complex document', () => {
        const markdown = `
# Main Header

**Bold text** and *italic text*.

1. First ordered item
2. Second ordered item

- First unordered item
- Second unordered item

> A blockquote

\`\`\`
Code block
\`\`\`

[Link](http://example.com)
![Image](http://example.com/image.png)
        `;
        expect(markdownToHtml(markdown)).toMatchSnapshot();
    });

    it('should handle nested images within links', () => {
        const markdown = '[![Image](http://example.com/image.png)](http://example.com)';
        const expectedHtml = '<p><a href="http://example.com"><img src="http://example.com/image.png" alt="Image"/></a></p>';
        expect(markdownToHtml(markdown)).toBe(expectedHtml);
    });

    it('should handle nested links within strong', () => {
        const markdown = '**[This is a link](http://example.com)**';
        const expectedHtml = '<p><strong><a href="http://example.com">This is a link</a></strong></p>';
        expect(markdownToHtml(markdown)).toBe(expectedHtml);
    });

    it('should handle nested images within strong', () => {
        const markdown = '**![Image](http://example.com/image.png)**';
        const expectedHtml = '<p><strong><img src="http://example.com/image.png" alt="Image"/></strong></p>';
        expect(markdownToHtml(markdown)).toBe(expectedHtml);
    });

});
