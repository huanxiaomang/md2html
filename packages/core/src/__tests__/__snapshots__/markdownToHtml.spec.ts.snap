// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`markdownToHtml > should match snapshot for complex document 1`] = `
"<h1>Main Header</h1><p><strong>Bold text</strong> and <em>italic text</em>.</p><ol><li>First ordered item</li><li>Second ordered item</li></ol><ul><li>First unordered item</li><li>Second unordered item</li></ul><blockquote><p>A blockquote</p></blockquote><pre><code>Code block
</code></pre><p><a href="http://example.com">Link</a></p><p><img src="http://example.com/image.png" alt="Image"/></p>"
`;

exports[`markdownToHtml > should match snapshot for mixed content 1`] = `"<h1>Header</h1><p><strong>Bold</strong> and <em>italic</em> text.</p><ul><li>List item 1</li><li>List item 2</li></ul><p><code>inline code</code></p>"`;
